File: problems/HackerRank/2521_exercises/linkedList/intermediate/mergeLists.c
SinglyLinkedListNode *mergeLists(SinglyLinkedListNode *head1, SinglyLinkedListNode *head2)
{

	if (head1 == NULL)
		return head2;
	if (head2 == NULL)
		return head1;

	struct SinglyLinkedListNode *dummyNode = (struct SinglyLinkedListNode *)malloc(sizeof(struct SinglyLinkedListNode));
	struct SinglyLinkedListNode *current = dummyNode;

	while (head1 && head2)
	{
		if (head1->data < head2->data)
		{
			current->next = head1;
			head1 = head1->next;
		}
		else
		{
			current->next = head2;
			head2 = head2->next;
		}
		current = current->next;
	}

	if (head1 != NULL)
	{
		current->next = head1;
	}
	else
	{
		current->next = head2;
	}

	SinglyLinkedListNode *mergedHead = dummyNode->next;
	free(dummyNode);

	return mergedHead;
}

// without using dummynode

SinglyLinkedListNode *mergeLists2(SinglyLinkedListNode *head1, SinglyLinkedListNode *head2)
{

	if (head1 == NULL)
		return head2;
	if (head2 == NULL)
		return head1;

	struct SinglyLinkedListNode *mergedHead = NULL;

	if (head1->data < head2->data)
	{
		mergedHead = head1;
		head1 = head1->next;
	}
	else
	{
		mergedHead = head2;
		head2 = head2->next;
	}

	struct SinglyLinkedListNode *current = mergedHead;

	while (head1 != NULL && head2 != NULL)
	{
		if (head1->data < head2->data)
		{
			current->next = head1;
			head1 = head1->next;
		}
		else
		{
			current->next = head2;
			head2 = head2->next;
		}

		current = current->next;
	}

	current->next = head1 != NULL ? head1 : head2;

	return mergedHead;
}

File: problems/HackerRank/week8/mergeLists.py
def newNode(data):
    n = SinglyLinkedListNode(data)
    n.next = None
    return n

def mergeLists(head1, head2):
    
    if not head1:
        return head2
    
    if not head2:
        return head1
    
    dummy = SinglyLinkedListNode(-1)
    dummy.next = None;
    curr = dummy
    
    while head1 and head2:
        if head1.data < head2.data:
            new = newNode(head1.data)
            head1 = head1.next
            curr.next = new
        else:
            new = newNode(head2.data)
            head2 = head2.next
            curr.next = new
            
        curr = curr.next
            
    while head1:
        new = newNode(head1.data)
        head1 = head1.next
        curr.next = new
        curr = curr.next
        
    while head2:
        new = newNode(head2.data)
        head2 = head2.next
        curr.next = new
        curr = curr.next
        
    return dummy.next


def mergeLists2(head1, head2):
    # Base cases
    if head1 is None:
        return head2
    if head2 is None:
        return head1

    # Pick the smaller head, recur on the rest
    if head1.data < head2.data:
        head1.next = mergeLists(head1.next, head2)
        return head1
    else:
        head2.next = mergeLists(head1, head2.next)
        return head2